app/src/main/java/com/example/dogbook/main/fragments/PostDetailsFragment.java

```diff
@@ -71,6 +77,44 @@ private void initView(View view) {
         rvPostDetails.setLayoutManager(layoutManager);
         refreshComments();
 
+        btnComment = view.findViewById(R.id.btnComment);
+        etComposeComment = view.findViewById(R.id.etComposeComment);
+        btnComment.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                uploadComment();
+            }
+        });
+    }
+
+    private void uploadComment() {
+        String commentText = etComposeComment.getText().toString();
+        if (commentText.length() == 0) {
+            Toast.makeText(getContext(), "Your comment is empty!", Toast.LENGTH_SHORT).show();
+            return;
+        }
+        Comment comment = new Comment();
+        comment.setAuthor(ParseUser.getCurrentUser());
+        comment.setPost(post);
+        comment.setContent(commentText);
+        comment.saveInBackground(new SaveCallback() {
+            @Override
+            public void done(ParseException e) {
+                if (e == null) {
+                    comments.add(comment);
+                    adapter.notifyItemInserted(comments.size());
+                    etComposeComment.setText("");
+                    //Disable and enable edit text to hide keyboard
+                    etComposeComment.setEnabled(false);
+                    etComposeComment.setEnabled(true);
+                    rvPostDetails.smoothScrollToPosition(comments.size());
+                    Log.i(TAG, "Comment uploaded");
+                    return;
+                }
+                Toast.makeText(getContext(), "Could not upload comment", Toast.LENGTH_SHORT).show();
+                Log.e(TAG, "Issue uploading comment", e);
+            }
+        });
```

Comment:
I think we could pull a lot of this out so that the posting comment functionality lives with your other data methods and then you just implement a callback here that lets you handle what you do on failure and success